db/main/migrate/20180509000000_add_active_flag_to_crons.rb

```diff
@@ -0,0 +1,13 @@
+class AddActiveFlagToCrons < ActiveRecord::Migration[4.2]
+  self.disable_ddl_transaction!
+
+  def change
+    change_table :crons do |t|
+      t.boolean :active
+    end
+
+    change_column_default(:crons, :active, true)
```

Comment:
this is a potentially unsafe operation as per [Safe Operations For High Volume PostgreSQL](https://www.braintreepayments.com/blog/safe-operations-for-high-volume-postgresql/). given that crons is a relatively small table it may be fine.

but it's something we need to watch out for when running the migration, and potentially cancel if it takes more than a couple seconds (since it will completely lock the table). I would recommend running the alter table command manually in order to have more control over this. by running it `heroku pg:psql` we can use ctrl+c to cancel it -- which afaik is not possible when running it through migrations.

Reply:
@igorwwwwwwwwwwwwwwwwwwww 
I probably misunderstood that post. I thought, setting the default for a column should be a separate command in a migration. That makes it safe for high volume dbs. Is that incorrect?

There are 30620 crons on org and 4776 crons on com. 

On com, the table size seems small for adding a column to. WDYT?

To further my own learning: how do you even estimate what is a large table size for a migration? Or, how do you even estimate how long a migration will take to run? 🤔

Reply:
I double checked, and you are absolutely right, adding a default value to a nullable column is indeed a safe operation. :) It is only in the case of non-nullable columns where this is an issue.

One thing worth noting though is that in that case the existing data in the table will not be updated. It will be null everywhere. So to set the value on existing rows we would need to run `UPDATE table SET active = true WHERE id BETWEEN x and y` in batches. But we'd need to make sure those run in a separate transaction each which I believe `disable_ddl_transaction!` will accomplish.

As a side-note: estimating how long it will take is kind of difficult. Number of rows and table size on disk can be good indicators. The tables are quite small. org is 4 megs which looks like it might be ok:

```
➜  ~ heroku pg:table-size -a travis-production | grep crons
 crons                | 3872 kB
➜  ~ heroku pg:table-size -a travis-pro-production | grep crons
 crons                     | 888 kB
```

But the best is still to try it imo. :) From this test is looks like adding the column with a default value should take less than a second on org: https://gist.github.com/igorwwwwwwwwwwwwwwwwwwww/87551e8a8ddf3a089ada751354c93835

Given that ^ it may be easier to try adding the column with a default value than having to fiddle with messy workarounds, as you suggested. I think it's worth a shot! :)

I'd try running it on production with `statement_timeout` set, that way, if it takes too long, it will cancel automatically. You could run something like this on psql console:

```
set statement_timeout to 2000; -- 2 seconds is the maximum time we will accept
alter table crons add column active boolean default true;
```

So yeah, I'd say let's try that! 👍

Reply:
Thanks for the detailed explanation @igorwwwwwwwwwwwwwwwwwwww 
I am going with the earlier version of the migration that alters the table and set a default value, all at once. ~230ms is quick enough!